File: singlyLinkedList.c
#include<stdio.h>
#include<stdlib.h>

typedef struct node{
    int data;
    struct node *next;
}node;

typedef struct list{
    node *head;
}list;

//function declarations
void init(list *l);
void insertBeginning(list *l,int data);
void insertEnd(list *l,int data);
void deleteBeginning(list *l);
void deleteEnd(list *l);
void insertAt(list *l,int data,int pos);
void deleteAt(list *l,int pos);
void reverse(list *l);
void disp(list *l);
void clear(list *l);

    void init(list *l)
{
    l->head = NULL;
}

void insertBeginning(list *l, int data){
    node *p;
    p = (node*)malloc(sizeof(node));
    if(p==NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    p->data = data;
    p->next = l->head;
    l->head = p;
    printf("%d Inserted successfully\n", data);
    return;
}

void insertEnd(list *l, int data){
    node *p,*q;
    p = (node*)malloc(sizeof(node));
    if(p==NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    p->data = data;
    p->next = NULL;
    q = l->head;
    if(q==NULL){
        l->head = p;
        printf("%d inserted successfully\n", data);
        return;
    }
    while(q->next != NULL){
        q = q->next;
    }
    q->next = p;
    printf("%d inserted successfully\n", data);
    return;
}

void deleteBeginning(list *l){
    node *p,*q;
    p=l->head;
    if(p==NULL){
        printf("List is empty\n");
        return;
    }
    q=p->next;
    l->head = q;
    printf("%d deleted successfully\n",p->data);
    free(p);
    return;
}

void deleteEnd(list *l){
    node *p,*q;
    p=l->head;
    if(p==NULL){
        printf("List is empty\n");
        return;
    }
    q=p->next;
    if(q==NULL){
        l->head = NULL;
        printf("%d deleted successfully\n", p->data);
        free(p);
        return;
    }
    while(q->next != NULL){
        p=q;
        q = q->next;
    }
    p->next = NULL;
    printf("%d deleted successfully\n", q->data);
    free(q);
    return;
}

void insertAt(list *l, int data, int pos){
    if(pos < 1){
        printf("Invalid position\n");
        return;
    }
    int length = 0;
    node *p,*q,*r;
    r = (node*)malloc(sizeof(node));
    if (r == NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    r->data = data;
    p = l->head;
    while(p!=NULL){
        length++;
        p=p->next;
    }
    if(pos > length+1){
        printf("Invalid position\n");
        return;
    }
    p = l->head;
    if(pos == 1){
        r->next = p;
        l->head = r;
        printf("%d inserted successfully\n", data);
        return;
    }
    int i=1;
    while(i!=pos){
        q = p;
        p = p->next;
        i++;
    }
    q->next = r;
    r->next = p;
    printf("%d inserted successfully\n", data);
    return;
}

void deleteAt(list *l, int pos){
    if (pos < 1){
        printf("Invalid position\n");
        return;
    }
    if (l->head == NULL){
        printf("List is empty\n");
        return;
    }
    int length = 0;
    node *p, *q;
    p = l->head;
    while (p != NULL){
        length++;
        p = p->next;
    }
    if (pos > length){
        printf("Invalid position\n");
        return;
    }
    p = l->head;
    if (pos == 1){
        l->head = p->next;
        printf("%d deleted successfully\n",p->data);
        free(p);
        return;
    }
    int i=1;
    while (i != pos){
        q = p;
        p = p->next;
        i++;
    }
    q->next = p->next;
    printf("%d deleted successfully\n", p->data);
    free(p);
    return;
}

void reverse(list *l){
    if(l->head == NULL){
        printf("List is empty\n");
        return;
    }
    if(l->head->next == NULL){
        printf("List reversed\n");
        return;
    }
    node *p,*q,*r;
    r = l->head;
    q = r->next;
    p = q->next;
    r->next = NULL;
    while(p != NULL){
        q->next = r;
        r=q;
        q=p;
        p=p->next;
    }
    q->next = r;
    l->head = q;
    printf("List reversed\n");
}

void disp(list *l){
    if(l->head == NULL){
        printf("List is empty\n");
        return;
    }
    node *p = l->head;
    printf("List contains: ");
    while(p!=NULL){
        printf("%d ",p->data);
        p = p->next;
    }
    printf("\n");
    return;
}

void clear(list *l){
    node *p,*q;
    p=l->head;
    while(p!=NULL){
        q = p;
        p = p->next;
        free(q);
    }
    l->head = NULL;
    printf("List cleared\n");
    return;
}

void main(){
    int loop=1;
    int choice,data,pos;
    list list;
    init(&list);
    printf("\n\t-------------------\n\tSINGLY LINKED LIST\n\t-------------------\n");
    while(loop){
        printf("\n----------------------------\n1. Insert in the beginning\n2. Insert in the end\n3. Delete from the beginning\n4. Delete from the end\n5. Insert at\n6. Delete at\n7. Reverse the list\n8. Display\n9. Clear\n10. Exit\n");
        printf("\nSelect an option: ");
        scanf("%d",&choice);
        printf("\n");
        switch(choice){
            case 1: //insert beginning
                printf("Enter element: ");
                scanf("%d",&data);
                insertBeginning(&list,data);
                break;

            case 2: //insert end
                printf("Enter element: ");
                scanf("%d", &data);
                insertEnd(&list, data);
                break;
            
            case 3: //delete beginning
                deleteBeginning(&list);
                break;
            
            case 4: //delete end
                deleteEnd(&list);
                break;
            
            case 5: //insert at
                printf("Enter element: ");
                scanf("%d",&data);
                printf("Enter position: ");
                scanf("%d",&pos);
                insertAt(&list,data,pos);
                break;
            
            case 6: //delete at
                printf("Enter position: ");
                scanf("%d", &pos);
                deleteAt(&list,pos);
                break;
            
            case 7: //reverse
                reverse(&list);
                break;
            
            case 8: //display
                disp(&list);
                break;

            case 9: //clear
                clear(&list);
                break;

            case 10: //exit
                clear(&list);
                loop = 0;
                printf("------------EXIT------------\n");
                break;

            default:
                printf("Enter a valid option\n");
                break;        
                
        }
    }
}

File: Go/SinglyLinkedList/singlyLinkedList.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type List struct {
	head *Node
	length int
}


func initList() *List {
	var list *List = &List{}
	list.head = nil
	list.length = 0
	return list
}


func (list *List) insertBeginning(data int) {
	var newNode *Node = &Node{data: data, next: list.head}
	list.head = newNode
	list.length += 1
	fmt.Println(data, "Inserted successfully")
	return
}


func (list *List) insertEnd(data int) {
	var newNode *Node = &Node{data: data, next: nil}
	if (list.length == 0) {
		list.head = newNode
	} else {
		var p *Node = list.head
		for (p.next != nil) {
			p = p.next
		}
		p.next = newNode
	}	
	list.length += 1
	fmt.Println(data, "Inserted successfully")
	return
}


func (list *List) deleteBeginning() {
	if (list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return
	}

	var p *Node = list.head
	list.head = list.head.next
	list.length -= 1

	fmt.Println(p.data,"Deleted successfully")
	p = nil	// marked for the GC
	return 
}


func (list *List) deleteEnd() {
	if(list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return
	}

	if(list.length == 1) {
		var p *Node = list.head
		list.head = nil
		list.length -= 1
		fmt.Println(p.data,"Deleted successfully")
		p = nil // marked for the GC
		return	
	}

	var p *Node = list.head
	q := p
	for (p.next != nil) {
		q = p
		p = p.next
	}
	q.next = nil
	list.length -= 1

	fmt.Println(p.data,"Deleted successfully")
	p = nil // marked for the GC
	return
}


func (list *List) insertAt(data int, pos int) {
	if(pos < 0 || pos > list.length) {
		fmt.Println("Invalid position to insert at")
		return
	}

	if(pos == 0) {
		list.insertBeginning(data)
		return
	}

	if(pos == list.length) {
		list.insertEnd(data)
		return
	}

	var p *Node = list.head
	for i := 0; i < pos-1; i++ {
		p = p.next
	}
	var newNode *Node = &Node{data: data, next: nil}
	newNode.next = p.next
	p.next = newNode
	list.length += 1
	fmt.Println(data, "Inserted successfully")
	return
}


func (list *List) deleteAt(pos int) {
	if(pos < 0 || pos > list.length) {
		fmt.Println("Invalid position to delete at")
		return
	}

	if(pos == 0) {
		list.deleteBeginning()
		return
	}

	if(pos == list.length) {
		list.deleteEnd()
		return
	}

	var p *Node = list.head
	for i := 0; i < pos-1; i++ {
		p = p.next
	}

	q := p.next
	p.next = q.next
	list.length -= 1
	fmt.Println(q.data, "Deleted successfully")
	q = nil
	return
}


func (list *List) reverse() {
	if (list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return		
	}

	if (list.length == 1) {
		fmt.Print("Reversed | ")
		list.display()
		return
	}

	r := list.head
	q := r.next
	p := q.next

	r.next = nil
	for (p != nil) {
		q.next = r
		r = q
		q = p
		p = p.next
	}

	q.next = r
	list.head = q
	fmt.Print("Reversed | ")
	list.display()
	return
}


func (list *List) display() {
	var p *Node = list.head
	fmt.Print("Linked List: ")
	for (p != nil) {
		fmt.Print(p.data," -> ")
		p = p.next
	}
	fmt.Print("nil | Length: ", list.length,"\n")
	return
}

func (list *List) clear() {
	p := list.head
	q := list.head
	for (p != nil) {
		q = p.next
		p = nil // mark for cleanup by GC
		p = q
	}

	// reset
	list.head = nil
	list.length = 0
	fmt.Println("Linked list cleared")
	return
}

func main() {
	var list *List = initList()
	fmt.Println("Initial Length: ", list.length)

    list.display();

    list.insertEnd(4);
    list.insertBeginning(1);
    list.insertAt(2,1);
    list.insertAt(3,2);

    list.display();
    list.reverse();

    list.clear();

    list.insertEnd(4);
    list.insertBeginning(1);
    list.insertAt(2,1);
    list.insertAt(3,2);

    list.display();

    list.deleteAt(10);
    list.deleteBeginning();
    list.reverse();
    list.deleteEnd();

    list.deleteAt(0);
    list.deleteAt(0);

    list.reverse();
    list.clear();
}